lucene/core/src/java/org/apache/lucene/index/FieldUpdatesBuffer.java

```diff
@@ -0,0 +1,281 @@
+  private FixedBitSet hasValues;
+  private String[] fields;
+  private final boolean isNumeric;
+
+  private FieldUpdatesBuffer(Counter bytesUsed, DocValuesUpdate initialValue, int docUpTo, boolean isNumeric) {
+    this.bytesUsed = bytesUsed;
+    this.bytesUsed.addAndGet(SELF_SHALLOW_SIZE);
+    termValues = new BytesRefArray(bytesUsed);
+    termValues.append(initialValue.term.bytes);
+    fields = new String[] {initialValue.term.field};
+    bytesUsed.addAndGet(sizeOfString(initialValue.term.field));
+    docsUpTo = new int[] {docUpTo};
+    if (initialValue.hasValue == false) {
+      hasValues = new FixedBitSet(1);
+      bytesUsed.addAndGet(hasValues.ramBytesUsed());
+    }
+    this.isNumeric = isNumeric;
+    byteValues = isNumeric ? null : new BytesRefArray(bytesUsed);
+  }
+
+  private long sizeOfString(String string) {
+    return STRING_SHALLOW_SIZE + (string.length() * Character.BYTES);
+  }
+
+  FieldUpdatesBuffer(Counter bytesUsed, DocValuesUpdate.NumericDocValuesUpdate initialValue, int docUpTo) {
+    this(bytesUsed, initialValue, docUpTo, true);
+    if (initialValue.hasValue()) {
+      numericValues = new long[] {initialValue.getValue()};
+    } else {
+      numericValues = new long[] {0};
+    }
+    bytesUsed.addAndGet(Long.BYTES);
+  }
+
+  FieldUpdatesBuffer(Counter bytesUsed, DocValuesUpdate.BinaryDocValuesUpdate initialValue, int docUpTo) {
+    this(bytesUsed, initialValue, docUpTo, false);
+    if (initialValue.hasValue()) {
+      byteValues.append(initialValue.getValue());
+    }
+  }
+
+  void add(String field, int docUpTo, int ord, boolean hasValue) {
+    if (fields[0].equals(field) == false || fields.length != 1 ) {
+      if (fields.length <= ord) {
+        String[] array = ArrayUtil.grow(fields, ord+1);
+        if (fields.length == 1) {
+          Arrays.fill(array, 1, ord, fields[0]);
+        }
+        bytesUsed.addAndGet((array.length - fields.length) * RamUsageEstimator.NUM_BYTES_OBJECT_REF);
+        fields = array;
+      }
+      if (field != fields[0]) { // that's an easy win of not accounting if there is an outlier
+        bytesUsed.addAndGet(sizeOfString(field));
+      }
+      fields[ord] = field;
+    }
+
+    if (docsUpTo[0] != docUpTo || docsUpTo.length != 1) {
+      if (docsUpTo.length <= ord) {
+        int[] array = ArrayUtil.grow(docsUpTo, ord+1);
+        if (docsUpTo.length == 1) {
+          Arrays.fill(array, 1, ord, docsUpTo[0]);
+        }
+        bytesUsed.addAndGet((array.length-docsUpTo.length) * Integer.BYTES);
+        docsUpTo = array;
+      }
+      docsUpTo[ord] = docUpTo;
+    }
+
+    if (hasValue == false || hasValues != null) {
+      if (hasValues == null) {
+        hasValues = new FixedBitSet(ord+1);
+        hasValues.set(0, ord);
+        bytesUsed.addAndGet(hasValues.ramBytesUsed());
+      } else {
+        FixedBitSet fixedBitSet = FixedBitSet.ensureCapacity(hasValues, ord+1);
```

Comment:
You might want to do something like this instead. Otherwise this is going to reuse the long[] but create a new FixedBitSet wrapper for every new update.

```java
if (fixedBitSet.length() <= ord) {
  FixedBitSet fixedBitSet = FixedBitSet.ensureCapacity(hasValues, ArrayUtil.oversize(ord + 1, 1));
  bytesUsed.addAndGet(fixedBitSet.ramBytesUsed()-hasValues.ramBytesUsed());
  hasValues = fixedBitSet;
}
```